### ISSUE TITLE
No sound devices detected

### ISSUE THREAD

**Issue by berlinbikerx** [eofster/Telephone#179]
I'm on a Mac Pro (late 2013) with USB Headset plugged in. I use Telephone compiled from latest master (SHA 323d3a3).

Telephone does not detect any sound devices (neither input, nor output, nor ring tone). By stepping through the code with the debugger, it throws an execption in function `firstOfDevices()` in file `FirstSystemAudioDevice.swift` when evaluating the predicate.

I can use the application, the sound is routed via the USB headset (which is set as default device in system settings), but I cannot choose other devices.

suspected function:

```swift
private func firstOfDevices(devices: [SystemAudioDevice], predicate: (SystemAudioDevice) -> Bool) throws -> SystemAudioDevice {
    /* exception here */
    if let result =  devices.filter(predicate).first {
        return result
    } else {
        throw Error.SystemAudioDeviceNotFoundError
    }
}
```

Call stack:

```
#0	0x0000000100437182 in (firstOfDevices in _586F29BB323C62219C48D74BD21F770F)([SystemAudioDevice], predicate : (SystemAudioDevice) -> Bool) throws -> SystemAudioDevice at Telephone/Domain/FirstSystemAudioDevice.swift:30
#1	0x0000000100436913 in FirstSystemAudioDevice.init(devices : [SystemAudioDevice], predicate : (SystemAudioDevice) -> Bool) throws -> FirstSystemAudioDevice at Telephone/Domain/FirstSystemAudioDevice.swift:22
#2	0x000000010043339d in FirstBuiltInSystemSoundIO.init(devices : [SystemAudioDevice]) throws -> FirstBuiltInSystemSoundIO at Telephone/Domain/FirstBuiltInSystemSoundIO.swift:23
#3	0x0000000100473719 in SelectedSystemSoundIO.init(devices : SystemAudioDevices, userDefaults : UserDefaults) throws -> SelectedSystemSoundIO at Telephone/UseCases/SelectedSystemSoundIO.swift:31
#4	0x000000010046c660 in UserAgentSoundIOSelectionInteractor.(updateSoundIO in _74A537D9711B0007DEFCFB8327CA65BB)() throws -> () at Telephone/UseCases/UserAgentSoundIOSelectionInteractor.swift:54
#5	0x000000010046b679 in UserAgentSoundIOSelectionInteractor.execute() throws -> () at Telephone/UseCases/UserAgentSoundIOSelectionInteractor.swift:40
#6	0x000000010046f70f in protocol witness for ThrowingInteractor.execute() throws -> () in conformance UserAgentSoundIOSelectionInteractor at Telephone/UseCases/UserAgentSoundIOSelectionInteractor.swift:37
#7	0x0000000100054b62 in UserAgentSoundIOSelector.selectUserAgentSoundIO(UserAgent) throws -> () at Telephone/Telephone/UserAgentSoundIOSelector.swift:28
#8	0x0000000100054d15 in UserAgentSoundIOSelector.userAgentDidFinishStarting(UserAgent) -> () at Telephone/Telephone/UserAgentSoundIOSelector.swift:35
#9	0x0000000100054f3c in protocol witness for UserAgentEventTarget.userAgentDidFinishStarting(UserAgent) -> () in conformance UserAgentSoundIOSelector at Telephone/Telephone/UserAgentSoundIOSelector.swift:33
#10	0x000000010006dec6 in UserAgentNotificationsToEventTargetAdapter.(SIPUserAgentDidFinishStarting in _291F899CB9FD03FC4BA1CB5ECE15B4A3)(NSNotification) -> () at Telephone/Telephone/UserAgentNotificationsToEventTargetAdapter.swift:65
```

**Comment by berlinbikerx** [comment]
Some further investigation: It's not the predicate that throws, but the function itself, because there is no builtin input device (the Mac Pro does not have a builtin mic).

When I plug in a mic into the headphones port (e.g. earplugs from an iPhone), everything is fine.

**Comment by eofster** [comment]
Thank you for the feedback! #157 is related to this problem.

**Comment by eofster** [comment]
> I can use the application, the sound is routed via the USB headset (which is set as default device in system settings), but I cannot choose other devices.

What other devices do you have configured? And how exactly does it look that you can’t choose other devices? They are not shown in settings?

**Comment by berlinbikerx** [comment]
The standard configurations are:

Output (Ausgabe):
- internal speakers
- line out

![sound-outputs](https://cloud.githubusercontent.com/assets/8158500/13994589/cfa78a50-f125-11e5-8a46-89202dd490fe.png)

Input (Eingabe)
(no device)

![sound-inputs](https://cloud.githubusercontent.com/assets/8158500/13994607/e76d66aa-f125-11e5-87c6-11aed7b3a919.png)

which results in completely empty drop down boxes in Telephone's sound settings:

![telephone-sound-settings](https://cloud.githubusercontent.com/assets/8158500/13994620/fcc5153e-f125-11e5-9a20-d877d248a1d4.png)

As soon as I plug in a input device, the drop downs are populated.

I guess the processing is aborted completely due to the exception which is thrown when no input device is found.

**Comment by eofster** [comment]
@berlinbikerx could you please try again with the latest master? The app now should not abort completely when sound IO is not found. The preferences should now be able to display either input or output alone. The underlying SIP library should now use its notion of default sound device when no sound device is present in the system. And I wonder if it doesn’t crash in this case. I can’t test it myself because my computer does have built-in devices.

**Comment by eofster** [comment]
Should be fixed in #157.

**Comment by berlinbikerx** [comment]
With the latest master, everything works as expected. With no devices plugged in, there is no input device available and for the output there are the integrated speakers and line out available.

Thanks for the quick fix.

![bildschirmfoto 2016-04-05 um 11 38 26](https://cloud.githubusercontent.com/assets/8158500/14277316/03046d12-fb23-11e5-97ed-3aee92a349e1.png)
